Khalil/5732 Adjust Blocktime Controller

kc1116 opened the pull request:
This PR adjusts the Blocktime controller and adds processing for the EpochCommittedPhaseStarted and EpochExtended events, which replace the EpochSetupPhaseStarted and EpochEmergencyFallbackTriggered event handlers. The EpochEmergencyFallbackTriggered and EpochSetupPhaseStarted events are no longer needed because the same information can be derived from the EpochExtended and EpochCommittedPhaseStarted events, respectively.
Additionally, now that we can enter and exit the happy paths and fallback paths of the state machine, it is important to process epoch-related events in the order they are received. This PR consolidates the previous multiple channels for processing epoch-related events into a single channel.

AlexHentschel commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -531,6 +531,18 @@ func (ctl *BlockTimeController) EpochEmergencyFallbackTriggered() {
+	ctl.nextEpochFinalView = nil
+}
+
+// EpochRecovered handles the epoch recovered protocol event.
+func (ctl *BlockTimeController) EpochRecovered(nextEpochFinalView *uint64) {
+	ctl.nextEpochFinalView = nextEpochFinalView
+}
```
I would recommend to pass `nextEpochFinalView` by value. Internally, the block time controller represents the `nextEpochFinalView` as a pointer, because there might be no next epoch specified atm (e.g. during the staking phase). However, this is the controllers _internal_ way of representing all possible cases.

In contrast, `EpochRecovered` specifically means that the subsequent epoch has now been specified. In this context, a nil value for `nextEpochFinalView` is incompatible with the protocol.

AlexHentschel reviewed (changes requested):
I hope it would be apparent that the current modifications of the [`BlockTimeController`](https://github.com/onflow/flow-go/blob/c8eb32fa3f97e3c3f2034d34d0d78af20b3a2a37/consensus/hotstuff/cruisectl/block_time_controller.go) are breaking concurrency safety. I leave it as an exercise to work out where and how.

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -462,16 +448,35 @@ func (ctl *BlockTimeController) measureViewDuration(tb TimedBlock) error {
+		// set epochFallbackTriggered to true when we encounter the first extension
+		ctl.epochFallbackTriggered = true
+	}
+	ctl.curEpochFinalView = extension.FinalView
```
I would suggest adding sanity checks on these view values. There is something very similar [in the Consensus Committee](https://github.com/onflow/flow-go/pull/6154/files#diff-7236d580d88b376c2040caf16dec5857f2c23366052f90e6f161d6d4646ca5b3R52-R59), I think we can use the same checks.

kc1116 replied:
https://github.com/onflow/flow-go/pull/6102/commits/cc6810ab137fe0f0aa5ece2f8d8272dab692e06c

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller_test.go:
```diff
@@ -232,50 +232,78 @@ func (bs *BlockTimeControllerSuite) TestInit_EpochFallbackTriggered() {
 	bs.AssertCorrectInitialization()
```
I think we can remove this test case (`TestInit_EpochFallbackTriggered`) because the component shouldn't need to behave differently when EFM is triggered. 

Ideally we would replace this test with one where we initialize the component when the current epoch already contains extensions. However, I'm not sure that we can usefully test that case at this level because we are mocking out the `protocol.Epoch`.

kc1116 replied:
https://github.com/onflow/flow-go/pull/6102/commits/a221aee99638fd22c68156f45e2ceff6e9f734ed , are there any tests that don't mock the `protocol.Epoch`?

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller_test.go:
```diff
@@ -232,50 +232,78 @@ func (bs *BlockTimeControllerSuite) TestInit_EpochFallbackTriggered() {
-func (bs *BlockTimeControllerSuite) TestEpochFallbackTriggered() {
+// TestOnEpochExtended ensures that the epoch configuration is updated when EpochExtended events are processed.
+func (bs *BlockTimeControllerSuite) TestOnEpochExtended() {
+	nextEpoch := mockprotocol.NewEpoch(bs.T())
+	nextEpoch.On("Counter").Return(bs.epochCounter+1, nil)
+	nextEpoch.On("FinalView").Return(bs.curEpochFinalView*2, nil)
+	nextEpoch.On("TargetDuration").Return(bs.EpochDurationSeconds(), nil)
+	nextEpoch.On("TargetEndTime").Return(bs.curEpochTargetEndTime+bs.EpochDurationSeconds(), nil)
+	bs.epochs.Add(nextEpoch)
```
What's the reasoning for adding a `nextEpoch` here?

kc1116 replied:
Good catch https://github.com/onflow/flow-go/pull/6102/commits/f2d65065d7f9883aca707699d96be8bd8a3a9cc3

jordanschalm reviewed (commented):
Summarizing main suggestions below 👇 

### Event Handling
The changes to event processing cause the worker thread to spin when no work is available. See [this comment](https://github.com/onflow/flow-go/pull/6102#discussion_r1670954630)

### Initializing in an epoch with extensions
This PR implements processing epoch extensions while running, but I don't think it will correctly handle the case where the component is initialized and the current epoch already contains extensions. 

In `initEpochInfo`, we initialize the epoch state at startup using the latest finalized state. 
https://github.com/onflow/flow-go/blob/77528dd08c1989bf459bce0d7b295b18d2f3ee4e/consensus/hotstuff/cruisectl/block_time_controller.go#L171-L182
However, some of these values that we read from the finalized state **do not account for extensions** ⚠️ . 

I encountered a similar problem in https://github.com/onflow/flow-go/pull/6154 and addressed it by modifying the `FinalView` method of `protocol.Epoch` to account for any extensions. 
https://github.com/onflow/flow-go/blob/0b1efe2400628d1f2ab6a7f8fa5999c2fcb7d97d/state/protocol/inmem/epoch.go#L77-L82

I think we should apply the same strategy to all methods of `protocol.Epoch` that can be affected by epoch extensions: `TargetEndTime`, `TargetDuration`.

The logic for `TargetEndTime` is straightforward. The logic for `TargetDuration` is trickier (I suspect computing `TargetDuration` here will get us much of the toward https://github.com/onflow/flow-go/issues/6194). 

### Change to EpochCommitted events
Since we changed from processing EpochSetup to EpochCommit events in this PR, we should adjust the corresponding initialization logic to only include next epoch data when we are in the EpochCommitted phase:
https://github.com/onflow/flow-go/blob/77528dd08c1989bf459bce0d7b295b18d2f3ee4e/consensus/hotstuff/cruisectl/block_time_controller.go#L187

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -462,16 +448,35 @@ func (ctl *BlockTimeController) measureViewDuration(tb TimedBlock) error {
+		ctl.epochFallbackTriggered = true
+	}
+	ctl.curEpochFinalView = extension.FinalView
+	ctl.curEpochTargetEndTime = extension.TargetEndTime
```
Just spoke with @durkmurder about this as he's planning to pick up #6194. We agreed that removing `TargetEndTime` from the extension model and moving that logic into the `protocol.Epoch` API method makes sense (reduces some duplication). That would require us to query `state.AtBlockID(refBlock).Epochs().Current().TargetEndTime()` here, rather than reading directly from the extension.

kc1116 replied:
Sounds good

kc1116 replied:
https://github.com/onflow/flow-go/pull/6102/commits/1f605be6ed39d809328e69e789eb489f79dbd538

kc1116 commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -273,57 +281,29 @@ func (ctl *BlockTimeController) processEventsWorkerLogic(ctx irrecoverable.Signa
 				ctx.Throw(err)
 				return
 			}
+		default:
```
https://github.com/onflow/flow-go/pull/6102/commits/f8440429ab4e52bec455ed6336aa7be442ef496b

kc1116 commented:
@jordanschalm relevant changes 

- reverted changes to worker processing func https://github.com/onflow/flow-go/pull/6102/commits/f8440429ab4e52bec455ed6336aa7be442ef496b
- updated branch with Epochs API changes that take EpochExtensions into account for API methods https://github.com/onflow/flow-go/pull/6102/commits/7be849dabb04f5c223e65f979605ca4dcfd9d349
- Change init epoch info phase check to `if phase == flow.EpochPhaseCommitted` https://github.com/onflow/flow-go/pull/6102/commits/e16a282e51cd1c429acd1a1bd165d86513d9c6f8

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -450,17 +430,25 @@ func (ctl *BlockTimeController) measureViewDuration(tb TimedBlock) error {
-	if !ctl.epochFallbackTriggered {
-		// set epochFallbackTriggered to true when we encounter the first extension
-		ctl.epochFallbackTriggered = true
+func (ctl *BlockTimeController) processEpochExtended(header *flow.Header, extension flow.EpochExtension) error {
```
```suggestion
func (ctl *BlockTimeController) processEpochExtended(first *flow.Header) error {
	finalView, err := ctl.state.AtHeight(header.Height).Epochs().Current().FinalView()
	if err != nil {
		return fmt.Errorf("failed to retrieve final view at block id %s", header.ID())
	}
	targetEndTime, err := ctl.state.AtHeight(header.Height).Epochs().Current().TargetEndTime()
	if err != nil {
		return fmt.Errorf("failed to retrieve target end time at block id %s", header.ID())
	}
```

A few small suggestions here:
- since we need to query the protocol state anyway, get both the final view and the target end time from querying the protocol state to eliminate the possibility that the two parameters are inconsistent
- use the `first` parameter name to match the interface definition and the other event handler function in this component
- use AtHeight (save an ID computation and match the other event handler function in this component)

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -450,17 +430,25 @@ func (ctl *BlockTimeController) measureViewDuration(tb TimedBlock) error {
+	if err != nil {
+		return fmt.Errorf("failed to retrieve target end time at block id %s", header.ID())
+	}
+	ctl.curEpochTargetEndTime = targetEndTime
```
```suggestion
	ctl.curEpochTargetEndTime = targetEndTime
	ctl.curEpochTargetDuration = targetDuration
```
We also need to update target duration here. We can get this value from the protocol state as well.
```go
ctl.state.AtHeight(header.Height).Epochs().Current().TargetDuration()
```

Let's also update [the test](https://github.com/onflow/flow-go/blob/a221aee99638fd22c68156f45e2ceff6e9f734ed/consensus/hotstuff/cruisectl/block_time_controller_test.go#L253) to check that this value is updated

kc1116 replied:
https://github.com/onflow/flow-go/pull/6102/commits/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad

AlexHentschel commented on consensus/hotstuff/cruisectl/block_time_controller_test.go:
```diff
@@ -223,59 +216,89 @@ func (bs *BlockTimeControllerSuite) TestInit_EpochSetupPhase() {
-		header := unittest.BlockHeaderFixture(unittest.HeaderWithView(bs.initialView + 1 + uint64(i)))
-		header.ParentID = unittest.IdentifierFixture()
-		bs.ctl.OnBlockIncorporated(model.BlockFromFlow(header))
+	assert.Equal(bs.T(), extensionTargetTime, bs.ctl.curEpochTargetEndTime)
```
can we also check the duration here? Just for good measure, so we have all possibly variable quantities checked here:
```suggestion
	assert.Equal(bs.T(), extensionTargetTime, bs.ctl.curEpochTargetEndTime)
	assert.Equal(bs.T(), extensionTargetDuration, bs.ctl.curEpochTargetDuration)
```

AlexHentschel commented on consensus/hotstuff/cruisectl/block_time_controller_test.go:
```diff
@@ -223,59 +216,89 @@ func (bs *BlockTimeControllerSuite) TestInit_EpochSetupPhase() {
+	}, time.Second, time.Millisecond)
+
+	assert.Equal(bs.T(), bs.curEpochFinalView*2, *bs.ctl.nextEpochFinalView)
+	assert.Equal(bs.T(), bs.curEpochTargetEndTime+bs.EpochDurationSeconds(), *bs.ctl.nextEpochTargetEndTime)
```
Could we also check that `bs.ctl.curEpochTargetDuration` has the expected value?

kc1116 replied:
https://github.com/onflow/flow-go/pull/6102/commits/c1bb33e73346e1922c5ab69916b340b971b84e93

AlexHentschel commented on consensus/hotstuff/cruisectl/block_time_controller_test.go:
```diff
@@ -223,59 +216,89 @@ func (bs *BlockTimeControllerSuite) TestInit_EpochSetupPhase() {
-	now = time.Now().UTC()
-	assert.Equal(bs.T(), now.Add(bs.config.FallbackProposalDelay.Load()), bs.ctl.getProposalTiming().TargetPublicationTime(17, now, unittest.IdentifierFixture()))
+
+	assert.Equal(bs.T(), bs.curEpochFinalView*2, *bs.ctl.nextEpochFinalView)
```
Could we also check that `bs.ctl.curEpochTargetDuration` has the expected value?

AlexHentschel reviewed (approved):
Very nice. Especially like that in your tests, you are verifying that the controller tolerates duplicates 👏

AlexHentschel commented:
I noticed that we duplicate notable amounts of code:
* checks that the `epochInfo` values are not nil in the [initialization](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L335-L344) and epoch [transition](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L335-L344)
* handling errors from querying the values in the [initialization](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L150-L172) and [update](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L477-L488)

I feel we could consolidate this 
1. by splitting the [`BlockTimeController.epochInfo`](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L79-L80) into the current and next epoch:
   ```golang
   currentEpochTiming epochTiming
   nextEpochTiming    *epochTiming
   ```
2. the struct `epochTiming` would then just describe a single epoch:
    ```golang
    // epochTiming encapsulates the timing information of one specific epoch:
    type epochTiming struct {
    	curEpochFirstView      uint64 // first view of the epoch's view range
    	curEpochFinalView      uint64 // last view of the epoch's view range
    	curEpochTargetDuration uint64 // desired total duration of the epoch in seconds
    	curEpochTargetEndTime  uint64 // target end time of the epoch, represented as Unix Time [seconds]
    }

    // newEpochTiming queries the timing information from the given `epoch` and returns it as a new `epochTiming` instance.
    func newEpochTiming(epoch protocol.Epoch) (*epochTiming, error) {
    	curEpochFirstView, err := epoch.FirstView()
    	if err != nil {
    		return nil, fmt.Errorf("could not retrieve epoch's first view: %w", err)
    	}
    	curEpochFinalView, err := epoch.FinalView()
    	if err != nil {
    		return nil, fmt.Errorf("could not retrieve epoch's final view: %w", err)
    	}
    	curEpochTargetDuration, err := epoch.TargetDuration()
    	if err != nil {
    		return nil, fmt.Errorf("could not retrieve epoch's target duration: %w", err)
    	}
    	curEpochTargetEndTime, err := epoch.TargetEndTime()
    	if err != nil {
    		return nil, fmt.Errorf("could not retrieve epoch's target end time: %w", err)
    	}

    	return &epochTiming{
    		curEpochFirstView:      curEpochFirstView,
    		curEpochFinalView:      curEpochFinalView,
    		curEpochTargetDuration: curEpochTargetDuration,
    		curEpochTargetEndTime:  curEpochTargetEndTime,
    	}, nil
    }

    // isFollowedBy determines whether nextEpoch is indeed the direct successor of the receiver,
    // based on the view ranges of both epochs.
    func (et *epochTiming) isFollowedBy(nextEpoch *epochTiming) bool {
    	return et.curEpochFinalView+1 == nextEpoch.curEpochFirstView
    }
    ```
3. I would really like a sanity check:
   * here, we just assume that we haven't missed anyting and there are no bugs anywhere that could compromise the correctness: https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L350
   * However, each epoch specifies its first view, so we can actually check. 

   Therefore, I included the method `epochTiming.isFollowedBy`. With the sanity check the method [`BlockTimeController.initEpochInfo`](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L146C33-L146C46) would then look like (a lot more concise in my opinion): 
    ```golang
    // initEpochInfo initializes the epochInfo state upon component startup.
    // No errors are expected during normal operation.
    func (ctl *BlockTimeController) initEpochInfo() error {
    	finalSnapshot := ctl.state.Final()

    	currentEpochTiming, err := newEpochTiming(finalSnapshot.Epochs().Current())
    	if err != nil {
    		return fmt.Errorf("failed to retrieve the current epoch's timing information: %w", err)
    	}
    	ctl.currentEpochTiming = *currentEpochTiming

    	phase, err := finalSnapshot.EpochPhase()
    	if err != nil {
    		return fmt.Errorf("could not check snapshot phase: %w", err)
    	}
    	if phase == flow.EpochPhaseCommitted {
    		ctl.nextEpochTiming, err = newEpochTiming(finalSnapshot.Epochs().Next())
    		if err != nil {
    			return fmt.Errorf("failed to retrieve the next epoch's timing information: %w", err)
    		}
    		if !currentEpochTiming.isFollowedBy(ctl.nextEpochTiming) {
    			return fmt.Errorf("failed to retrieve the next epoch's timing information: %w", err)
    		}
    	}

    	return nil
    }
    ```
4. similarly for [`processEpochCommittedPhaseStarted`](https://github.com/onflow/flow-go/blob/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad/consensus/hotstuff/cruisectl/block_time_controller.go#L474):
    would also include the sanity check

kc1116 commented:
Refactor epochInfo into currentEpochTiming and *nextEpochTiming https://github.com/onflow/flow-go/pull/6102/commits/84fb5a20eb255d197ed7d2baccf1627dc5f596cd

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller.go:
```diff
@@ -328,32 +332,28 @@ func (ctl *BlockTimeController) processIncorporatedBlock(tb TimedBlock) error {
-	ctl.nextEpochFinalView = nil
-	ctl.nextEpochTargetDuration = nil
-	ctl.nextEpochTargetEndTime = nil
+	ctl.currentEpochTiming.firstView = ctl.nextEpochTiming.firstView
+	ctl.currentEpochTiming.finalView = ctl.nextEpochTiming.finalView
+	ctl.currentEpochTiming.targetDuration = ctl.nextEpochTiming.targetDuration
+	ctl.currentEpochTiming.targetEndTime = ctl.nextEpochTiming.targetEndTime
```
```suggestion
	ctl.currentEpochTiming = *ctl.nextEpochTiming
```
Optionally, a more concise way to move the data over.

jordanschalm commented on consensus/hotstuff/cruisectl/block_time_controller_test.go:
```diff
@@ -220,27 +222,41 @@ func (bs *BlockTimeControllerSuite) TestOnEpochExtended() {
+
 	bs.ctl.EpochExtended(bs.epochCounter, header, extension)
 
 	// expect epoch fallback triggered to be set to true when first extension is encountered indicating we are in epoch fallback mode
```
```suggestion
	// Check component state after the epochEvents channel is empty, indicating the event has been processed. 
```

jordanschalm reviewed (approved):
Great work 🎸
